R/methSeg.R

```diff
@@ -88,126 +195,119 @@
   
-  args.fastseg[["x"]]=obj
-  
-  # do the segmentation
-  #seg.res=fastseg(obj)
-  seg.res <- do.call("fastseg", args.fastseg)
-  #seg.res <- do.call("fastseg2", args.fastseg)
-  
-  # stop if segmentation produced only one range
-  if(length(seg.res)==1) {
-    warning("segmentation produced only one range, no mixture modeling possible.")
-    seg.res$seg.group <- "1"
-    return(seg.res)
-  }
-  
-  # if joining, do not show first clustering
-  if(join.neighbours) {
-    diagnostic.plot.old = diagnostic.plot
-    diagnostic.plot = FALSE
+  if(cores==1){
+    
+    ##coerce object to granges
+    if(class(obj)=="methylRaw" | class(obj)=="methylRawDB") {
+      obj= as(obj,"GRanges")
+      ## calculate methylation score 
+      mcols(obj)$meth=100*obj$numCs/obj$coverage
+      ## select only required mcol
+      obj = obj[,"meth"]
+    }else if (class(obj)=="methylDiff" | class(obj)=="methylDiffDB") {
+      obj = as(obj,"GRanges")
+      ## use methylation difference as score
+      obj = obj[,"meth.diff"]
+    }else if (class(obj) != "GRanges"){
+      stop("only methylRaw or methylDiff objects ", 
+           "or GRanges objects can be used in this function")
+    }
+    
+    # destrand
+    strand(obj) <- "*"
+    
+    # Run segmentation
+    seg.res = .run.fastseg(obj)
   }
   
-  if("initialization" %in% names(args.Mclust)){
-    if("subset" %in% names(args.Mclust[["initialization"]])) {
-      if(length(args.Mclust[["initialization"]][["subset"]]) < 9 ){
-        stop("too few samples, increase the size of subset.") 
+  if(cores>1){
+    
+    ## Non-tabix files  
+    if (class(obj)=="methylDiff" | class(obj)=="methylRaw" | 
+        class(obj)=="GRanges"){
+      
+      ##coerce object to granges
+      if(class(obj)=="methylRaw") {
+        obj= as(obj,"GRanges")
+        ## calculate methylation score 
+        mcols(obj)$meth=100*obj$numCs/obj$coverage
+        ## select only required mcol
+        obj = obj[,"meth"]
+      } else if(class(obj)=="methylDiff") {
+        obj = as(obj,"GRanges")
+        ## use methylation difference as score
+        obj = obj[,"meth.diff"]
+      }else if (class(obj) != "GRanges"){
+        stop("only methylRaw or methylDiff objects ", 
+             "or GRanges objects can be used in this function")
+      }
+      # destrand
+      strand(obj) <- "*"
+      
+      chrs = seqlevels(obj)
+      # Split the input object based on chromosome
+      # run methSeg on chromosomes (Could be in parallel or not)
+      seg.res.list <- mclapply(chrs, function(chr){
+        seg.res = .run.fastseg(obj[seqnames(obj)==chr])
+      }, mc.cores=cores)
+      
+      # Merge resulting segments again to whole genome.
+      # suppressWarnings -> combined objs dont have chrs in common
+      seg.res = suppressWarnings( do.call("c", seg.res.list) )
+      
+      ## Tabix files
+    } else if(class(obj)=="methylDiffDB" | class(obj)=="methylRawDB"){
+      
+      .run.fastseg.tabix = function(gr0, ...){
```

Comment:
I would suggest a function which actually takes the class of the object as argument:
```
.run.fastseg.tabix = function(gr0, class ,...) {

### and then you can directly set the colnames
.setMethylDBNames(df2getcolnames,class)
}
```